src/app/content/highlights/components/SummaryPopup/utils.ts

```diff
@@ -0,0 +1,33 @@
+import { Highlight } from '@openstax/highlighter/dist/api';
+import queryString from 'query-string';
+import React from 'react';
+import { useSelector } from 'react-redux';
+import { book as bookSelector } from '../../../selectors';
+import { Book, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../../types';
+import { findArchiveTreeNode } from '../../../utils/archiveTreeUtils';
+import { getBookPageUrlAndParams } from '../../../utils/urlUtils';
+
+const createHighlightLink = (
+  { id, anchor }: Highlight,
+  page?: LinkedArchiveTreeSection | LinkedArchiveTreeNode,
+  book?: Book
+) => {
+  const search = queryString.stringify({ target: JSON.stringify({ id, type: 'highlight' }) });
+  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${search}#${anchor}`;
+};
+
+export const useCreateHighlightLink = (highlight: Highlight) => {
+  const [link, setLink] = React.useState('');
+  const book = useSelector(bookSelector);
+
+  const page = React.useMemo(() => book
+    ? findArchiveTreeNode(book.tree, highlight.sourceId)
+    : undefined
+  , [book, highlight.sourceId]);
+
+  React.useEffect(() => {
+    setLink(createHighlightLink(highlight, page, book));
+  }, [highlight, page, book]);
+
+  return link;
+};
```

Comment:
Doesn't this essentaily double the amount of re-renders for whatever component would need it? First the component would rerender, then the effect would run, then it would rerender again because the state is set here too.

Reducing amount of rerenders is not necessarily a priority but i feel like this is also against the principle of only "one source of truth"

Besides that, I'm not sure if highly specialized hooks are a good practice. I feel like there should be only a function that gets memoized, unless it's not expensive. I don't really have an another argument against one time use hooks so maybe my opinion is invalid.

Reply:
I feel like `useSelector` is doing almost the same thing, but in this case rewriting it to the `React.useMemo` made it 3 times smaller so I guess this should be our approach

Reply:
but why are you using state in this case anyway? highlight, page, and book will all be set from the start so i'm not seeing why any change tracking is necessary

Reply:
Only the current page is set from the start, but we need `Page` for specific `highlight`. I wanted to memoize results because in order to create a `linktToHighlight` we have to obtain a page from the book which then is passed to `getBookPageUrlAndParams` and doing so requires flattening of book tree and there is no reason to do this for each highlight on every rerender. Initially, I've used `hook` approach, but `useMemo` is okay in this case

Reply:
using state is not necessary, this is just something which I'm used to do since in the other project I'm writing quite a few hooks for async methods and then using a state is the only option